src/Orleans.Core/Runtime/AsynchAgent.cs

```diff
@@ -170,14 +173,36 @@ public bool CheckHealth(DateTime lastCheckTime)
         internal static bool IsStarting { get; set; }
 
         protected virtual ThreadPoolExecutorOptions.Builder ExecutorOptionsBuilder =>
-            new ThreadPoolExecutorOptions.Builder(Name, GetType(), Cts, loggerFactory, ExecutorFaultHandler);
+            new ThreadPoolExecutorOptions.Builder(Name, GetType(), Cts, loggerFactory).WithExecutionFilters(executorFaultHandler);
 
-        protected void ExecutorFaultHandler(Exception ex)
+        private sealed class ExecutorFaultHandler : ExecutionFilter
+        {
+            private readonly AsynchAgent agent;
+
+            public ExecutorFaultHandler(AsynchAgent agent)
+            {
+                this.agent = agent;
+            }
+
+            public override Func<Exception, Threading.ExecutionContext, bool> ExceptionHandler => (ex, context) =>
```

Comment:
can avoid an alloc here by moving this to the ctor and using a get-only property instead
![image](https://user-images.githubusercontent.com/203839/34801550-6aaedd98-f6bd-11e7-8ee2-7ebcece55e8a.png)

Reply:
At least the compiler avoided re-creating the closure over `agent` each time

Reply:
Allocations removed.